### PR TITLE
Sidecar translations

### PR DESCRIPTION
**Pull request by elia**
### Summary

Having sidecar translations is mentioned in multiple places in the project and the two apps using components I'm aware of are both doing it. Locally we solved with a less refined version of the attached code.

### Other Information

This builds on the work started in #74.

Expects a component to have a file with the same basename and the `.yml` extension, holding translations scoped to the component's path.

A component YML should look like this:

```yml
  en:
    hello: "Hello World!"
```

And inside the component will be possible to call

```erb
<%=  t(".hello") %>
```

Internally the scope will be expanded using the template's path.
As an example, the translation above from within `app/components/greeting/component.html.erb` will result in an internal `I18n` key like this:

```yml
  en:
    greeting:
      component:
        hello: "Hello World!"
```

---

_This PR is sponsored by @nebulab_

### CODE REVIEW

**Comment by elia** [comment]
For some reason Rails 5.0.0 on Ruby 2.7 was failing with a weird bundler error, updating to the latest patch seems to have fixed the problem.

**Comment by joelhawksley** [comment]
@elia thank you for the PR! We should be able to review it early next week. 

In the meantime, I believe we can revert the dependency version changes. We see that build failure from time to time, but haven't been able to identify the cause. Rebuilding usually fixes it.

**Comment by elia** [comment]
> @elia thank you for the PR! We should be able to review it early next week.

Awesome!

> In the meantime, I believe we can revert the dependency version changes. We see that build failure from time to time, but haven't been able to identify the cause. Rebuilding usually fixes it.

Done ✅ 

Do you want me to send [that change](https://github.com/elia/view_component/commit/e53b398) into another PR? Staying up-to-date seems a good way to avoid needless headaches.

File: lib/view_component/i18n.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    # As an example, the translation above from within `app/components/greeting/component.html.erb`
+    # will result in an internal I18n key like this:
+    #
+    #   en:
+    #     greeting:
+    #       component:
+    #         hello: "Hello World!"
+    #
+    #
+    def self.load(components_root: "#{Rails.root}/app/components", glob: "**/*component.{yml,yaml}")
```
Comment (BlakeWilliams) [inline lib/view_component/i18n.rb:30]: I really love this change, but I think it'd be great if we could make this part of the compilation step.

I think that'd let us more tightly couple components to translation files, removing the need to configure translations. I *think* it might also help with template reloading, but that may need more investigation.

Comment (BlakeWilliams) [reply lib/view_component/i18n.rb:30]: Building on that thought a bit, it'd be great if we could use the component's file name to determine the translation file name instead of requiring the translation files to end in `component.*`.

e.g. If I have `modal.rb`, I'd expect the translation file to be called `modal.yml`,  or `modal/modal.yml` (or `.yaml` as an extension).

**Review by BlakeWilliams — COMMENTED** [review]
Thanks for taking this on! This is a really great feature.

File: lib/view_component/i18n.rb (hunk 1 above)
Comment (elia) [reply lib/view_component/i18n.rb:30]: The idea of using the compilation step to add paths to I18n is really neat, and `source_location` is perfect to do that. 

Also about reloading I missed an initializer during the extraction from our application, which would be integrated in `engine.rb` (it's modeled after the original I18n reloading code):

```rb
# Reload component translations whenever we change them.
Rails.application.configure do
  config.after_initialize do |app|
    app.config.file_watcher.new [], "#{Rails.root}/app/components" => ".yml" do
      I18n.reload!
    end.tap do |reloader|
      app.reloaders << reloader
      app.reloader.to_run do
        reloader.execute_if_updated { require_unload_lock! }
      end
    end
  end
end
```

With an hypothetical `modal.rb` component it should be already possible to adjust the glob, even to a generic `**/*.y{a,}ml`, to encompass any YAML in the components folder.

---

So, before proceeding with the implementation inside `compile`, I'd like to have your opinion on a couple of considerations:
- I foresee possible headaches for applications not being strict about only using the translations within the component
- We might still need the reloader to properly pickup changes (renames/deletions) 

The code would go more or less like this:

```rb
def compile
  unless I18n.paths.include? component_locale
    I18n.paths << component_locale
    I18n.reload!
  end
  # ...
end
```

File: lib/view_component/i18n.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+      Dir["#{components_root}/#{glob}"].each.with_object({}) do |path, translations|
+        relative_path = Pathname(path).relative_path_from(Pathname(components_root)).to_s
+        component_translations = YAML.load_file path
+        scopes = relative_path.sub(/\.ya?ml/, "").split("/")
+
+        component_translations.to_h.each do |locale, scoped_translations|
```
Comment (cj) [inline lib/view_component/i18n.rb:36]: I have done some testing, and to allow options like:

```yml
en:
  greeting:
    hello: "Hello World!"

# AND

en:
  greeting:
    component:
      hello: "Hello World!"
```

Doing:

```ruby
        component_translations.to_h.each do |locale, scoped_translations|
          translations[locale] ||= {}
          translations[locale].deep_merge!(scoped_translations)
        end
```

seems to work better.  Thoughts?

File: docs/index.md [hunk 3]
```diff
@@ -1117,9 +1117,42 @@ app/components
...
+```yml
+en:
+  hello: "Hello world!"
+```
+
+This will be expanded internally to use the full path of the file as a scope:
+
+```yml
+en:
+  greeting:
```
Comment (jonspalmer) [inline docs/index.md:1141]: should this be `greeter` given the file paths?

File: lib/view_component/i18n.rb (hunk 1 above)
Comment (jonspalmer) [reply lib/view_component/i18n.rb:30]: > I foresee possible headaches for applications not being strict about only using the translations within the component

Can you share a bit more about that concern? What's an example of "not being strict"?

File: docs/index.md [hunk 4]
```diff
@@ -1117,9 +1117,42 @@ app/components
...
+
+In order to support `I18n` translations scoped to the component you can add a dynamic translations loader in `config/locale/components.rb`.
+
+```ruby
+require "view_component/i18n"
+ViewComponent::I18n.load components_root: "#{Rails.root}/app/components", glob: "**/*component.yml"
+```
+
+This allows for adding a "sidecar" YAML file that will define translations
+scoped to the component, e.g. `app/components/greeter/component.yml`:
```
Comment (jonspalmer) [inline docs/index.md:1130]:
```suggestion
scoped to the component, e.g. `app/components/greeting/component.yml`:
```

File: docs/index.md [hunk 5]
```diff
@@ -976,7 +976,7 @@ function importAll(r) {
   r.keys().forEach(r)
 }
 
-importAll(require.context("../components", true, /_component\.js$/))
+importAll(require.context("../components", true, /[_\/]component.js$/))
```
Comment (Spone) [inline docs/index.md:979]: I guess this unrelated change should be part of another PR.

Also, why removing the `\` in `component\.js`? It's useful to escape the dot that otherwise means "any character".

File: docs/index.md (hunk 4 above)
Comment (Spone) [reply docs/index.md:1130]: What about using `example` as the component name to stay consistent with the rest of the docs?

File: lib/view_component/i18n.rb [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    #   t(".hello") # => "Hello World!"
+    #
+    # Internally the scope will be expanded using the template's path.
+    # As an example, the translation above from within `app/components/greeting/component.html.erb`
+    # will result in an internal I18n key like this:
+    #
+    #   en:
+    #     greeting:
+    #       component:
+    #         hello: "Hello World!"
```
Comment (Spone) [inline lib/view_component/i18n.rb:27]: I'm curious about the rationale behind this naming/namespacing.

By default the naming of the template file for an `ExampleComponent` with sidecar would be: `app/components/example_component/example_component.html.erb` so I think the docs should show this case instead.

What would the structure of the I18n key be in this case? `en.example_component.example_component.hello`, right?

File: lib/view_component/i18n.rb (hunk 2 above)
Comment (elia) [reply lib/view_component/i18n.rb:36]: The idea is that a component would always use the relative notation, if that's the case the component's naming shouldn't matter, e.g. both these scenarios are expected to work:

---
`app/components/greeting.html.erb`
```erb
<%= t(".hello") %>
```
`app/components/greeting.yml`
```erb
en:
  hello: "Hello World!"
```
---
`app/components/greeting/component.html.erb`
```erb
<%= t(".hello") %>
```
`app/components/greeting/component.yml`
```erb
en:
  hello: "Hello World!"
```
---
Am I missing some context with regard to this?

File: lib/view_component/i18n.rb (hunk 6 above)
Comment (Spone) [reply lib/view_component/i18n.rb:27]: :wave: @elia did you see this comment?

Comment (elia) [reply lib/view_component/i18n.rb:27]: _Sorry, I did read it but then somehow skipped it when replying_

I'll update the doc, the use of `example/component.*` was just coming from the way we organized them in the app I extracted this from.

**Comment by elia** [comment]
The CI is having random failures with bundler (v1) and `Browser did not produce websocket url within 5 seconds (Ferrum::ProcessTimeoutError)` in the PVC job, I don't have the permission to restart the build, let me know if I should just keep trying using fake force-pushes.

**Comment by Spone** [comment]
> The CI is having random failures with bundler (v1) and `Browser did not produce websocket url within 5 seconds (Ferrum::ProcessTimeoutError)` in the PVC job, I don't have the permission to restart the build, let me know if I should just keep trying using fake force-pushes.

I retried it a couple times, `CI / pvc (pull_request)` is failing every time. Not sure what to do @joelhawksley?

File: lib/view_component/i18n.rb (hunk 1 above)
Comment (BlakeWilliams) [reply lib/view_component/i18n.rb:31]: > What I mean is that translations are expected to be there globally, but if we load them only when the component is compiled they might not be there in every environment/situation (think background jobs, rake tasks, partials, etc.) unless we ensure components are all eagerly compiled all the time.

Translations being available globally breaks the implicit and explicit encapsulation that view component assumes. For example, if we have two components with the same name they might have an i18n naming collision. I think it's important to handle that edge case.

File: lib/view_component/i18n.rb (hunk 6 above)
Comment (Spone) [reply lib/view_component/i18n.rb]: *Don't worry about it ;)*

I find the `en.example_component.example_component.hello` I18n key not ideal, and I'm wondering if it shouldn't be something like `en.view_component.example_component.hello` or some other naming that:

- uses a `view_component` or `components` namespace to prevent collision with other parts of the app
- removes the component name duplication
- is consistent whether you have:
    - a *regular* component (with a template placed in `app/components/example_component.html.erb`)
    - a *sidecar* component (with a template placed in `app/components/example_component/example_component.html.erb`)
    - an *inline* component (so the `I18n.t` method is called from the component `call` method)

What do you think?

**Comment by joelhawksley** [comment]
> > The CI is having random failures with bundler (v1) and `Browser did not produce websocket url within 5 seconds (Ferrum::ProcessTimeoutError)` in the PVC job, I don't have the permission to restart the build, let me know if I should just keep trying using fake force-pushes.
> 
> I retried it a couple times, `CI / pvc (pull_request)` is failing every time. Not sure what to do @joelhawksley?

We've marked that build as optional while we investigate ❤️

**Comment by patchwork-her** [comment]
You finally did something. Can you admit to me please?

**Comment by patchwork-her** [comment]
Illegal activity if you ask me because I didn't authorize none of this and
I've been hacked out of it.

File: lib/view_component/i18n.rb (hunk 6 above)
Comment (elia) [reply lib/view_component/i18n.rb]: I like the idea of both the prefix and collapsing the sidecar components path with the duplicate name, the only thing I want to point out before trying to implement them is that we need to keep this configuration in synch with `virtual_path` in `ViewComponent::Base` because that's the one that will drive relative translations inside templates.

So, probably we should start moving some of this stuff to engine-level configurations, including maybe the components root and the glob for the YAML files. That would be ok?

Comment (Spone) [reply lib/view_component/i18n.rb]: I'd like to request @joelhawksley and @BlakeWilliams opinion about this, as I'm not sure what's the best move.

Comment (BlakeWilliams) [reply lib/view_component/i18n.rb]: Building on https://github.com/github/view_component/pull/552#discussion_r584988143, if we go down that path I think we could avoid coupling i18n key names to specific components because each component will have access to the global i81n, but also its sidecar i18n, but not the sidecar i18n of other components. 

> So, probably we should start moving some of this stuff to engine-level configurations, including maybe the components root and the glob for the YAML files. That would be ok?

We should avoid configuration wherever possible (or at least have some great defaults), so if we can make i18n work out of the box that would be preferable.

Comment (elia) [reply lib/view_component/i18n.rb]: @BlakeWilliams please checkout #660 which features a version of sidecar i18n as you intended, the global one is still available via `I18n.t` or `helpers.t`.

**Comment by elia** [comment]
This has been replaced by #660 in which the I18n lookup will only include the sidecar translations instead of having them mixed with the global Rails I18n.

**Comment by patchwork-her** [comment]
My "one" big thing I truly pay for is this. I need it
